backend/webdav/webdav.go

```diff
@@ -1154,19 +1163,146 @@ func (o *Object) Update(ctx context.Context, in io.Reader, src fs.ObjectInfo, op
+		ContentType:   contentType,
+		Options:       options,
+		ExtraHeaders:  extraHeaders,
+	}
+
+	err = o.fs.pacer.CallNoRetry(func() (bool, error) {
+		resp, err = o.fs.srv.Call(ctx, &opts)
+		return o.fs.shouldRetry(resp, err)
+	})
+	if err != nil {
+		// Give the WebDAV server a chance to get its internal state in order after the
+		// error.  The error may have been local in which case we closed the connection.
+		// The server may still be dealing with it for a moment. A sleep isn't ideal but I
+		// haven't been able to think of a better method to find out if the server has
+		// finished - ncw
+		time.Sleep(1 * time.Second)
+		// Remove failed upload
+		_ = o.Remove(ctx)
+		return err
+	}
+	return nil
+
+}
+
+
+// Chunked update for Nextcloud (see https://docs.nextcloud.com/server/20/developer_manual/client_apis/WebDAV/chunking.html)
+func (o *Object) _UpdateChunked(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (err error) {
+
+	var uploadDir string = strings.ReplaceAll(path.Join(o.fs.root,o.remote), "/", "_")
+	
+	saveRoot := o.fs.root
+	saveRemote := o.remote
+	o.fs.root = "/"
+	o.fs.srv.SetRoot(strings.Replace(o.fs.endpointURL,"files","uploads",1))
+	
+	err = o.fs.mkdir(ctx, uploadDir)
+	if err != nil {
+		// Delete upload dir
+		rmdirerr := o.fs.Rmdir(ctx, uploadDir)
+		if rmdirerr != nil {
+			// FIXME return 2 errors
+			return errors.Wrap(rmdirerr, "Remove upload dir failed, Upload mkUploadDir failed")
+		}
+		return errors.Wrap(err, "Update mkUploadDir failed")
+	}
+
+	size := src.Size()
+	var uploadedSize int64 = 0
+	for uploadedSize < size {
+		o.fs.root = uploadDir
+		// Upload chunk
+		contentLength := o.fs.opt.NextcloudChunkSize
+		if size - uploadedSize < contentLength {
+			contentLength = size - uploadedSize
+		}
+		o.remote = fmt.Sprintf("%015d", uploadedSize)
+		chunkPath := o.filePath()
+		extraHeaders := map[string]string{}
+		err = o._UpdateSimple(ctx, io.LimitReader(in, o.fs.opt.NextcloudChunkSize), chunkPath, contentLength, "", extraHeaders, options...)
+		
+		if err != nil {
+			// Delete upload dir
+			o.fs.root = ""
+			rmdirerr := o.fs.Rmdir(ctx, uploadDir)
+			if rmdirerr != nil {
+				// FIXME return 2 errors
+				return errors.Wrap(rmdirerr, "Remove upload dir failed, Upload chunk failed")
+			}
+			return errors.Wrap(err, "Upload chunk failed")
+		}
+		uploadedSize += contentLength
+	}
+	 
+	// Finish
+	o.fs.root = saveRoot
+	o.remote = saveRemote
+	var resp *http.Response
+	opts := rest.Opts{
+		Method:        "MOVE",
+		Path:          rest.URLPathEscape(path.Join(uploadDir,".file")),
+		NoResponse:    true,
 		Options:       options,
 	}
+	opts.ExtraHeaders = map[string]string{}
+	destinationURL, err := rest.URLJoin(o.fs.endpoint, o.filePath())
+	if err != nil {
+		return errors.Wrap(err, "Finalize chunked upload couldn't join URL")
+	}
+	opts.ExtraHeaders["Destination"] = destinationURL.String()
 	if o.fs.useOCMtime || o.fs.hasMD5 || o.fs.hasSHA1 {
```

Comment:
This code block in the `if` is now duplicated. Factor it out?

Reply:
✅